stl/inc/algorithm

```diff
@@ -10204,12 +10204,24 @@ namespace ranges {
                 "The lower bound cannot be greater than the upper bound in a call to std::ranges::clamp "
                 "(N4861 [alg.clamp]/2).");
 
-            if (_STD invoke(_Pred, _STD invoke(_Proj, _Val), _STD invoke(_Proj, _Lo))) {
-                return _Lo;
-            }
+            if constexpr (!is_copy_constructible_v<decltype(_STD invoke(_Proj, _Val))>) {
+                // move-only types
+                if (_STD invoke(_Pred, _STD invoke(_Proj, _Val), _STD invoke(_Proj, _Lo))) {
+                    return _Lo;
+                }
 
-            if (_STD invoke(_Pred, _STD invoke(_Proj, _Hi), _STD invoke(_Proj, _Val))) {
-                return _Hi;
+                if (_STD invoke(_Pred, _STD invoke(_Proj, _Hi), _STD invoke(_Proj, _Val))) {
+                    return _Hi;
+                }
+            } else {
+                auto&& _Temp = _STD invoke(_Proj, _Val);
+                if (_STD invoke(_Pred, _Temp, _STD invoke(_Proj, _Lo))) {

```

Comment:
Both of those should be `std::forward<decltype(_Temp)>(_Temp)`, and the other branch of the `if constexpr` should just go away.

Reply:
> The double forward is safe because `regular_invocable` requires that the invocation of the predicate not modify `_Temp` in a manner observable to equality-preserving expressions.

As I read this, double forward is wrong. Could you help me to understand what did I miss?

There is a constraint `indirect_strict_weak_order<_Pred, projected<const _Ty*, _Pj>>`.

`indirect_strict_weak_order` requires:

1. `strict_weak_order<_Pred&, iter_value_t<I1>&, iter_value_t<I2>&>`
1. `strict_weak_order<_Pred&, iter_value_t<I1>&, iter_reference_t<I2>>`
1. `strict_weak_order<_Pred&, iter_reference_t<I1>, iter_value_t<I2>&>`
1. `strict_weak_order<_Pred&, iter_reference_t<I1>, iter_reference_t<I2>>`
1. `strict_weak_order<_Pred&, iter_common_reference_t<I1>, iter_common_reference_t<I2>>`

where `I1` = `I2` = `projected<const _Ty*, _Pj>`.

`strict_weak_order<_Pred&, T, U>` requires relation<_Pred, T, U> which requires:

1. `predicate<_Pred&, T, T>`
1. `predicate<_Pred&, U, U>`
1. `predicate<_Pred&, T, U>`
1. `predicate<_Pred&, U, T>`

`predicate<_Pred&, T, U>` requires `regular_invocable<_Pred&, T, U>` which requires 

```
requires(F&& f, Args&&... args) {
    invoke(std::forward<F>(f), std::forward<Args>(args)...); // required to be equality-preserving
};
```

Let's take for example following `Pred`, `Proj` and `X`

```
struct Pred {
    bool operator()(std::shared_ptr<int> x, std::shared_ptr<int> y)
    {
        assert(x && y);
        return *x < *y;
    }
};

struct X
{
    X(int v) : x(std::make_shared<int>(v)){}
    std::shared_ptr<int> x;
};

struct Proj {
    std::shared_ptr<int> operator()(const X& x) {
        return x.x;
    }
};
```

Let's look at the requirements of `indirect_strict_weak_order<Pred, projected<const X*, Proj>>`:

1. `I1` is same as `I2` and is `projected<const X*, Proj>`.
2. `iter_value_t<I1>&` is `std::shared_ptr<int>&`.
3. `iter_reference_t<I1>` is `std::shared_ptr<int>`.
4. `iter_common_reference_t<I1>` is `std::shared_ptr<int>`.

Therefore there are following requirements:
1. `strict_weak_order<Pred&, std::shared_ptr<int>&, std::shared_ptr<int>&>` from `strict_weak_order<_Pred&, iter_value_t<I1>&, iter_value_t<I2>&>`
1. `strict_weak_order<Pred&, std::shared_ptr<int>&, std::shared_ptr<int>>` from `strict_weak_order<_Pred&, iter_value_t<I1>&, iter_reference_t<I2>>`
1. `strict_weak_order<Pred&, std::shared_ptr<int>, std::shared_ptr<int>&>` from `strict_weak_order<_Pred&, iter_reference_t<I1>, iter_value_t<I2>&>`
1. `strict_weak_order<Pred&, std::shared_ptr<int>, std::shared_ptr<int>>` from `strict_weak_order<_Pred&, iter_reference_t<I1>, iter_reference_t<I2>>`
1. `strict_weak_order<Pred&, std::shared_ptr<int>, std::shared_ptr<int>>` from `strict_weak_order<Pred&, iter_common_reference_t<I1>, iter_common_reference_t<I2>>`

Or if we omit same requirements:
1. `strict_weak_order<Pred&, std::shared_ptr<int>&, std::shared_ptr<int>&>`
1. `strict_weak_order<Pred&, std::shared_ptr<int>&, std::shared_ptr<int>>`
1. `strict_weak_order<Pred&, std::shared_ptr<int>, std::shared_ptr<int>&>`
1. `strict_weak_order<Pred&, std::shared_ptr<int>, std::shared_ptr<int>>`

Let's substitute further:
1. `predicate<Pred&, std::shared_ptr<int>&, std::shared_ptr<int>&>`
1. `predicate<Pred&, std::shared_ptr<int>&, std::shared_ptr<int>>`
1. `predicate<Pred&, std::shared_ptr<int>, std::shared_ptr<int>&>`
1. `predicate<Pred&, std::shared_ptr<int>, std::shared_ptr<int>>`

`predicate<Pred&, T, U>` requires `regular_invocable<Pred&, T, U>`:
1. `regular_invocable<Pred&, std::shared_ptr<int>&, std::shared_ptr<int>&>`
1. `regular_invocable<Pred&, std::shared_ptr<int>&, std::shared_ptr<int>>`
1. `regular_invocable<Pred&, std::shared_ptr<int>, std::shared_ptr<int>&>`
1. `regular_invocable<Pred&, std::shared_ptr<int>, std::shared_ptr<int>>`

As you can see there is no requirement of `regular_invocable` with `std::shared_ptr<int>&&` as either argument. Therefore there is no requirement that call to `std::invoke(_Pred, std::forward<decltype(_Temp)>(_Temp), std::invoke(_Proj, _Lo))` shall be equality-preserving.

In GCC 11.1 `std::ranges::clamp` works with this example but proposed implementation fails at runtime.
[link to godbolt](https://godbolt.org/z/4o9xbT3T4)

Reply:
Substituting `F` = `Pred&` and `Args...` = `shared_ptr<int>, shared_ptr<int>` into the definition of `regular_invocable`

```cpp
requires(F&& f, Args&&... args) {
    invoke(std::forward<F>(f), std::forward<Args>(args)...); // required to be equality-preserving
};
```
produces:

```cpp
requires(Pred& f, shared_ptr<int>&& arg0, shared_ptr<int>&& arg1) {
    invoke(std::forward<Pred&>(f), std::forward<shared_ptr<int>>(arg0), std::forward<shared_ptr<int>>(arg1));
};
```
which fails with your example because the `invoke` expression modifies its arguments.

In other words, `regular_invocable` doesn't distinguish between `T` and `T&&`.

Reply:
Thank you for your reply!

Given that `_STD invoke(_Pred, _STD forward<decltype(_Temp)>(_Temp), _STD invoke(_Proj, _Lo))` should be equality-preserving and (as you say) can't modify `_Temp` in any way, what example could benefit from using `_STD forward<decltype(_Temp)>(_Temp)` here?

After re-reading [concepts.equality] I have following (maybe wrong :) ) interpretation of "equality-preserving".

> An expression is equality-preserving if, given equal inputs, the expression results in equal outputs. The inputs to an expression are the set of the expression's operands. The output of an expression is the expression's result and all operands modified by the expression.

I read it as "equality-preserving" expression *can* modify its inputs (and it counts towards expression's result). But *given equal inputs* it must produce equal outputs (and modify its inputs "equally").

Stability requirement is stated "absent any explicit intervening modification of those input objects". So my understanding is that it is not guaranteed that calling stable equality-preserving expression (that modify some of its inputs) multiple times with same inputs produce same result. For this to be true inputs of this calls should be equal (but they were modified).

For example, it is stated that "Expressions declared in a requires-expression in the library clauses are required to be equality-preserving, except for those annotated with the comment “not required to be equality-preserving.”". And "The library clauses use a notational convention to specify which expressions declared in a requires-expression modify which inputs: except where otherwise specified, an expression operand that is a non-constant lvalue or rvalue may be modified".

I read it as rvalues (for example `std::forward<Args>(args)...` in `std::regular_invocable` concept) is expected to be modified if not stated explicitly to be not modified. On the other hand "Operands that are constant lvalues or rvalues are required to not be modified".

Example of not equality-preserving callable given in [concepts.regularinvocable] is "random number generator" and not "every callable that takes arguments by value".

Let's look at concept `assignable_from`. It has following `requires` clause:

    requires(LHS lhs, RHS&& rhs) {
      { lhs = std::forward<RHS>(rhs) } -> same_as<LHS>;
    };

There is no comment that `lhs = std::forward<RHS>(rhs)` is "not required to be equality-preserving". Does it mean that `assignable_from` can't modify its `rhs`?

(non-regular) `invocable` concept on the other hand contains such comment:

    template<class F, class... Args>
    concept invocable = requires(F&& f, Args&&... args) {
        invoke(std::forward<F>(f), std::forward<Args>(args)...); // not required to be equality-preserving
    };

so it is permitted to return different outputs for equal inputs and to modify not only non-constant lvalue or rvalue but even constant lvalue and rvalue (for example, calling const callable object can modify that callable object `mutable` members).

Is this interpretation wrong?

Reply:
`regular_invocable` explicitly requires the `invoke` expression to *both* be equality-preserving *and* not modify either the function object or its arguments. Those are distinct requirements.

The need to forward arises from the fact that the predicate is *not* required to be invocable with `iter_reference_t<I>&, iter_reference_t<I>` (or the reverse).

Reply:
Thank you for your reply! (and patience :) )

> Those are distinct requirements.

Oh, this is what I missed. :(

> The need to forward arises from the fact that the predicate is *not* required to be invocable with `iter_reference_t<I>&, iter_reference_t<I>` (or the reverse).

Predicate is required to be invocable with `iter_common_reference_t<I>, iter_common_reference_t<I>` for `I = std::projected<const T*, Proj>` and `std::indirectly_readable<I>`. 

`iter_reference_t<I>&` is always convertible to `iter_common_reference_t<I>` in this case, isn't it?

Reply:
> `iter_reference_t<I>&` is always convertible to `iter_common_reference_t<I>` in this case, isn't it?

I don't think it is required to be, but in any event "convertible to" is insufficient. `invocable<F, X> && convertible_to<Y, X>` doesn't imply `invocable<F, Y>`.